AutoNode/passphrase.py

```diff
@@ -0,0 +1,109 @@
+import subprocess
+import os
+import base64
+
+import pexpect
+from cryptography.fernet import Fernet
+from cryptography.fernet import InvalidToken
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, hmac
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+from pyhmy import (
+    cli,
+    Typgpy
+)
+
+from .common import (
+    validator_config,
+    node_config,
+    log
+)
+from .exceptions import (
+    InvalidWalletPassphrase
+)
+
+
+def _get_harmony_pid():
+    try:
+        return subprocess.check_output(["pgrep", "harmony"], env=os.environ).strip()
+    except (subprocess.TimeoutExpired, subprocess.CalledProcessError):
+        return b'0'
+
+
+def _get_process_info(pid):
+    assert isinstance(pid, bytes)
+    if pid == b'0':
+        return pid
+    try:
+        raw_output = subprocess.check_output(["ls", "-ld", f"/proc/{pid.decode().strip()}"], env=os.environ).strip()
+        lst = raw_output.split(' '.encode())
+        return b''.join(lst[1:-1]) if len(lst) >= 3 else b'0'
+    except (subprocess.TimeoutExpired, subprocess.CalledProcessError):
+        return b'0'
+
+
+def _get_node_based_salt():
+    key = ''.join([str(x) for x in node_config["public-bls-keys"]]
+                  + [str(validator_config["validator-addr"]), str(validator_config["identity"])]).encode()
+    return hmac.HMAC(key, hashes.SHA256(), backend=default_backend()).finalize()
+
+
+def _derive_wallet_encryption_key():
+    """
+    Create the wallet encryption key based on:
+        PBKDF2HMAC(PID(harmony) + ProcessInfo(harmony), salt=HMAC(node_bls_public_keys, Validator_Addr, Validator_ID))
+    Where node_config* are all node_configs as a dict, except for encrypted values.
+
+    This means that the encryption key is only valid for when AutoNode has a harmony node process running.
+    """
+    pid = _get_harmony_pid()
+    proc_info = _get_process_info(pid)
+    kdf = PBKDF2HMAC(
+        algorithm=hashes.SHA256(),
+        length=32,
+        salt=_get_node_based_salt().strip(),
+        iterations=50,
+        backend=default_backend()
+    )
+    return base64.urlsafe_b64encode(kdf.derive(pid.strip() + proc_info.strip()))
+
+
+def encrypt_wallet_passphrase(passphrase):
+    """
+    Encrypt the given passphrase.
+
+    Returned string can be stored on disk, will be invalidated once harmony process is stopped.
```

Comment:
good question. 

First, PID for the harmony process is not logged anywhere on purpose so that it is **harder** to get this info.
Moreover, linux never logs PID. PID in systemd is lost after the restart, PID in the journal is from python PID. 

But should the PID be known, there are many cryptographic defences in place. Defence in layers is the best way to do any encryption. A 'salt' based on the BLS keys, validator addr, and ID is used. Moreover, the key of the encryption is a derivation from PID + command used to run the node is also used + time that the process started. Lastly, once a node a killed, the encrypted passphrase gets deleted.

The likelihood of getting ALL of this info & satisfying the HMAC (integrity hash) of the passphrase is unlikely. 

The big thing tho is that this encrypted passphrase is NEVER getting sent over the internet, so overall I would say its a safe scheme.